### PR TITLE
[JBPM-9129] Start process without triggering a start node

### PR DESCRIPTION
**Pull request by elguardian**
added new command to restart a process from where it left

### CODE REVIEW

**Comment by elguardian** [comment]
depends on https://github.com/kiegroup/droolsjbpm-knowledge/pull/438

File: drools-core/src/main/java/org/drools/core/base/WrappedStatefulKnowledgeSessionForRHS.java [hunk 1]
```diff
@@ -775,19 +775,9 @@ public SessionClock getSessionClock() {
...
-//
-//    public void guardRuleUnit(RuleUnit ruleUnit, Activation activation) {
-//        delegate.getRuleUnitExecutor().guardRuleUnit(ruleUnit, activation);
-//    }
-//
-//    public void guardRuleUnit(Class<? extends RuleUnit> ruleUnitClass, Activation activation) {
-//        delegate.getRuleUnitExecutor().guardRuleUnit(ruleUnitClass, activation);
-//    }
+    @Override
+    public ProcessInstance startProcessFromNodeIds(String processId, Map<String, Object> params, String... nodeInstancesIds) {
```
Comment (krisv) [inline drools-core/src/main/java/org/drools/core/base/WrappedStatefulKnowledgeSessionForRHS.java]: Still seems to refer to nodeInstanceIds but I think these are just nodeIds (node definition ids).

File: drools-core/src/main/java/org/drools/core/command/runtime/process/StartProcessFromNodeIdsCommand.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+import org.kie.internal.process.CorrelationAwareProcessRuntime;
+import org.kie.internal.process.CorrelationKey;
+
+@XmlRootElement
+@XmlAccessorType(XmlAccessType.NONE)
+public class StartProcessFromNodeIdsCommand extends StartProcessCommand implements CorrelationKeyCommand {
+
+    private static final long serialVersionUID = -6839934578407787510L;
+
+    @XmlElementWrapper(name = "node-instance-id-list")
```
Comment (krisv) [inline drools-core/src/main/java/org/drools/core/command/runtime/process/StartProcessFromNodeIdsCommand.java]: Same here, this class seem to refer to instance ids in many places, but they are definition ids I believe.

**Review by krisv — CHANGES_REQUESTED** [review]

File: drools-core/src/main/java/org/drools/core/base/WrappedStatefulKnowledgeSessionForRHS.java (hunk 1 above)
Comment (elguardian) [reply drools-core/src/main/java/org/drools/core/base/WrappedStatefulKnowledgeSessionForRHS.java]: done.

File: drools-core/src/main/java/org/drools/core/command/runtime/process/StartProcessFromNodeIdsCommand.java (hunk 2 above)
Comment (elguardian) [reply drools-core/src/main/java/org/drools/core/command/runtime/process/StartProcessFromNodeIdsCommand.java]: done

File: drools-core/src/main/java/org/drools/core/command/runtime/process/StartProcessFromNodeIdsCommand.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+                    result.append(", ");
+                }
+                result.append(entry.getKey());
+                result.append("=");
+                result.append(entry.getValue());
+            }
+        }
+        result.append("]);");
+        return result.toString();
+    }
```
Comment (gmunozfe) [inline drools-core/src/main/java/org/drools/core/command/runtime/process/StartProcessFromNodeIdsCommand.java]: This is the same as parent but class name, that could be resolved by reflection or tailored with a protected attribute

File: drools-ruleunit/src/main/java/org/drools/ruleunit/executor/RuleUnitSessionImpl.java [hunk 4]
```diff
@@ -1006,4 +1006,14 @@ public void deactivate() {
     public boolean tryDeactivate() {
         return session.tryDeactivate();
     }
+
+    @Override
+    public ProcessInstance startProcessFromNodeIds(String processId, Map<String, Object> params, String... nodeInstancesIds) {
```
Comment (gmunozfe) [inline drools-ruleunit/src/main/java/org/drools/ruleunit/executor/RuleUnitSessionImpl.java]: Same comment as Kris here to be coherent, use _nodeIds_

File: drools-templates/src/test/java/org/drools/template/parser/ExternalSheetListenerTest.java [hunk 5]
```diff
@@ -454,6 +454,11 @@ public void fireUntilHalt() {
                 public void fireUntilHalt(org.kie.api.runtime.rule.AgendaFilter agendaFilter) {
 
                 }
+
+                @Override
+                public ProcessInstance startProcessFromNodeIds(String processId, Map<String, Object> params, String... nodeInstancesIds) {
```
Comment (gmunozfe) [inline drools-templates/src/test/java/org/drools/template/parser/ExternalSheetListenerTest.java]: and same here

File: drools-core/src/main/java/org/drools/core/command/runtime/process/StartProcessFromNodeIdsCommand.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2010 Red Hat, Inc. and/or its affiliates.
```
Comment (afalhambra) [inline drools-core/src/main/java/org/drools/core/command/runtime/process/StartProcessFromNodeIdsCommand.java]: Should be 2020
```suggestion
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
```

File: drools-core/src/main/java/org/drools/core/command/runtime/process/StartProcessFromNodeIdsCommand.java (hunk 3 above)
Comment (afalhambra) [inline drools-core/src/main/java/org/drools/core/command/runtime/process/StartProcessFromNodeIdsCommand.java]: Wonder if logging `nodeIds` list here could add some value when debugging? wdyt?

**Review by afalhambra — CHANGES_REQUESTED** [review]
Small changes requested, but looks good to me

File: drools-core/src/main/java/org/drools/core/command/runtime/process/StartProcessFromNodeIdsCommand.java (hunk 6 above)
Comment (elguardian) [reply drools-core/src/main/java/org/drools/core/command/runtime/process/StartProcessFromNodeIdsCommand.java]: done

File: drools-core/src/main/java/org/drools/core/command/runtime/process/StartProcessFromNodeIdsCommand.java (hunk 3 above)
Comment (elguardian) [reply drools-core/src/main/java/org/drools/core/command/runtime/process/StartProcessFromNodeIdsCommand.java]: removed.

File: drools-ruleunit/src/main/java/org/drools/ruleunit/executor/RuleUnitSessionImpl.java (hunk 4 above)
Comment (elguardian) [reply drools-ruleunit/src/main/java/org/drools/ruleunit/executor/RuleUnitSessionImpl.java]: done

File: drools-templates/src/test/java/org/drools/template/parser/ExternalSheetListenerTest.java (hunk 5 above)
Comment (elguardian) [reply drools-templates/src/test/java/org/drools/template/parser/ExternalSheetListenerTest.java]: done

**Review by afalhambra — APPROVED** [review]
Nice work!

**Comment by mareknovotny** [comment]
@elguardian please fix compilation failure on CDB https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/KIE/job/master/job/pullrequest/job/drools-compile-downstream-build/1499/console
it was revealed in another PR FDB

**Pull request merged**